packages/vendure-plugin-sort-by-popularity/src/sort.service.ts

```diff
@@ -1,12 +1,180 @@
+      .andWhere('order.orderPlacedAt is NOT NULL')
+      .andWhere('product.deletedAt IS NULL')
+      .andWhere('productVariant.deletedAt IS NULL')
+      .andWhere('product.enabled')
+      .andWhere('productVariant.enabled')
+      .andWhere('order_channel.id = :id', { id: channleId })
+      .addGroupBy('product.id')
+      .addOrderBy('count', 'DESC')
+      .getRawMany();
+    const maxCount = groupedOrderItems[0].count;
+    const maxValue = 1000;
+    const productRepository = this.connection.getRepository(ctx, Product);
+    await productRepository.save(
+      groupedOrderItems.map((gols) => {
+        return {
+          id: gols.product_id,
+          customFields: {
+            popularityScore: Math.round((gols.count / maxCount) * maxValue),
+          },
+        };
+      })
+    );
+    await this.assignScoreValuesToCollections(ctx);
+    return { success: true };
+  }
+  async assignScoreValuesToCollections(ctx: RequestContext) {
+    const allCollectionsScores = await this.getEachCollectionsScore(ctx);
+    await this.addUpTheTreeAndSave(allCollectionsScores, ctx);
+  }
+
+  async getEachCollectionsScore(
+    ctx: RequestContext
+  ): Promise<{ id: string; score: number }[]> {
+    const collectionsRepo = this.connection.getRepository(ctx, Collection);
+    const productsRepo = this.connection.getRepository(ctx, Product);
+    const allCollectionIds = await collectionsRepo
+      .createQueryBuilder('collection')
+      .select('id')
+      .getRawMany();
+    const productScoreSums: { id: string; score: number }[] = [];
+    const variantsPartialInfoQuery = collectionsRepo
+      .createQueryBuilder('collection')
+      .select('collection.id')
+      .leftJoin('collection.productVariants', 'productVariant')
+      .addSelect('productVariant.productId');
+    const productSummingQuery = productsRepo
+      .createQueryBuilder('product')
+      .select('SUM(product.customFieldsPopularityscore) AS productScoreSum');
+    for (const col of allCollectionIds) {
+      const variantsPartialInfo = await variantsPartialInfoQuery
+        .where('collection.id= :id', { id: col.id })
+        .getRawMany();
+
+      const productIds = variantsPartialInfo
+        .filter((i) => i.productVariant_productId != null)
+        .map((i) => i.productVariant_productId);
+
+      const uniqueProductIds = [...new Set(productIds)];
+
+      const summedProductsValue = await productSummingQuery
+        .where('product.id IN (:...ids)', { ids: uniqueProductIds })
+        .getRawOne();
+      productScoreSums.push({
+        id: col.id,
+        score: summedProductsValue.productScoreSum,
+      });
+    }
+    await collectionsRepo.save(
+      productScoreSums.map((collection) => {
+        return {
+          id: collection.id,
+          customFields: {
+            popularityScore: collection.score ?? 0,
+          },
+        };
+      })
+    );
+    return productScoreSums;
+  }
+
+  async addUpTheTreeAndSave(
```

Comment:
Could also use some js doc, telling the developers reading this that this function calculates scores for parent collections based on their child collections